cartographer/common/thread_pool.h

```diff
@@ -45,15 +57,24 @@ class ThreadPool : public ThreadPoolInterface {
   ThreadPool(const ThreadPool&) = delete;
   ThreadPool& operator=(const ThreadPool&) = delete;
 
+  void NotifyReady(Task* task) EXCLUDES(mutex_) override;
+
+  // TODO(gaschler): Remove all uses.
   void Schedule(const std::function<void()>& work_item) override;
 
+  // If the returned weak pointer is expired, 'task' has certainly completed,
+  // so dependants no longer need to add it as a dependency.
+  std::weak_ptr<Task> ScheduleWhenReady(std::unique_ptr<Task> task)
+      EXCLUDES(mutex_);
+
  private:
   void DoWork();
 
   Mutex mutex_;
   bool running_ GUARDED_BY(mutex_) = true;
   std::vector<std::thread> pool_ GUARDED_BY(mutex_);
-  std::deque<std::function<void()>> work_queue_ GUARDED_BY(mutex_);
+  std::deque<std::shared_ptr<Task>> task_queue_ GUARDED_BY(mutex_);
+  std::map<Task*, std::shared_ptr<Task>> tasks_not_ready_ GUARDED_BY(mutex_);
```

Comment:
unordered_map? You do not iterate over it anywhere I think.

Reply:
Done